Added sensitivity analysis to Cantherm

alongd opened the pull request:
Now users can run a Cantherm job (`KineticsJob` or `PressureDependenceJob`) and see to which wells or transitions states the calculated rates are most sensitive to (and hopefully later improve them).
The output of the SA is saved into a `sensitivity` folder in the output directory with a text file tabulating all normalized sensitivity coefficients per reaction (in both direction if the network reaction is reversible), and figures visualizing the sensitivity coefficients per network reaction at all conditions defined for the sensitivity analysis.

Test, example, and some documentation also added.

goldmanm commented on documentation/source/users/cantherm/input.rst:
```diff
@@ -611,6 +611,14 @@ This is also acceptable::
 
     kinetics('H + C2H4 <=> C2H5')
 
+If a sensitivity analysis is desired, simply add the conditions at which to calculate sensitivity coefficients
```
it might be useful to add the syntax for sensitivity using pressureDependence method (in addition to the kinetics method used here). 

we might also want to describe what the output of sensitivity analysis looks like in the documentation (including the folder name, how the sensitivity coefficients are normalized, etc.)

goldmanm commented on rmgpy/cantherm/pdepTest.py:
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
```
It seems like the actual operation being tested is inside the `setUp` method and not within any particular test itself. 

These also seem more like functional tests than unit tests, since you are running an entire module and checking for file output. Maybe this file should explicitly mention that it is a functional test, and it should be specified with the attribute `@attr('functional')` above the class line like in `testGenerateReactions.py`.

goldmanm commented on .gitignore:
```diff
@@ -43,6 +43,7 @@ nbproject/*
 .coverage
 testing/*
 htmlcov/*
+rmgpy/test_data/Cantherm/tst1/*
```
It might be better to have a `tearDown` method that cleans these files instead of adding them to gitignore

goldmanm commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -43,6 +43,7 @@
 
 import rmgpy.quantity as quantity
 import rmgpy.constants as constants
+from rmgpy.molecule.draw import MoleculeDrawer, createNewSurface
```
I am a little confused by the functionality of this. Is drawing the PES optional? If it is, should we describe that in documentation? If not, how much time is consumed by this operation relative to calculating the TST equation? When processing many reactions, are separate PES files created or just one with everything drawn on it (which I have seen cantherm do before)?

alongd replied:
It's fairly fast, so I included it by default (will not run if Cairo isn't installed, though).
I ran two `Kinetics` calcs including PES draw and SA for one of them in just under 10 seconds on my PC.
I just verified that it saves different plots for each PES.

alongd commented:
@goldmanm, thanks for your comments!
I addressed them with fixups (and dropped the gitignore one, so force-pushed), let me know if you have additional input for me.

mjohnson541 commented on rmgpy/cantherm/sensitivity.py:
```diff
@@ -207,3 +209,176 @@ def plot(self):
+            self.perturb(entry)
+            self.job.execute(outputFile=None, plot=False, print_summary=False)  # run the perturbed job
+            self.unperturb(entry)
+            if isinstance(entry, TransitionState):
```
Is this done because E0 always needs to be positive to prevent negative values near zero?

alongd replied:
I removed this part, so now we use semi-normalized coefficients, and E0 isn't needed

alongd commented:
@cgrambow, thanks for your offline comments yesterday.

Since in the normalized sensitivity calculation (`dln(r)/dln(E0)`) I multiply by `E0` (which might be negative, or even zero...) I got wrong coefficients (both in sign and in magnitude).
I now modified the calculation to semi-normalized coefficients (`dln(r)/dE0`), like the thermo sensitivity in RMG.

Now the figures make more sense. Here's the example we discussed:
![image](https://user-images.githubusercontent.com/16158262/40787019-8f3e5f06-64ba-11e8-99cc-04e16cdc5d4a.png)
I get the following sensitivities with tunnelling:
```
  Sensitivity of network reaction reactant <=> product :
=====================================================================================
| Well or TS           | Temperature (K) | Pressure (bar) | Sensitivity coefficient |
|----------------------+-----------------+----------------+-------------------------|
| reactant             | 1000.0          |     1.00       | +1.46e-04               |
| product              | 1000.0          |     1.00       | -3.07e-06               |
| (TS) TS1             | 1000.0          |     1.00       | -1.43e-04               |
=====================================================================================
```
![image](https://user-images.githubusercontent.com/16158262/40790346-b440d434-64c2-11e8-8351-26c4d933c7e2.png)

cgrambow commented:
Awesome! Those make a lot more sense.

alongd commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -295,3 +298,435 @@ def plot(self, outputDirectory):
+
+        # Determine the rectangles taken up by each well
+        # We'll use this to merge columns safely so that wells don't overlap
+        wellRects = []
```
I believe so. Right, in complex networks we do have overlap and reading might be hard

alongd commented on rmgpy/cantherm/sensitivity.py:
```diff
@@ -0,0 +1,209 @@
+        self.r_sa_rates = {}
+        self.f_sa_coefficients = {}
+        self.r_sa_coefficients = {}
+        self.perturbation = quantity.Quantity(1, 'kcal/mol')
```
I also think this shouldn't be hard-coded.
I changed the Class, so this is now an input parameter. We also iterate several times using `try - except` while capturing errors, and slightly decreasing the perturbation for better network stability

alongd commented:
@mjohnson541, thanks for your comments. I think I addressed everything, changes are already squashed.
If there are no more comments, I think this can get merged once the tests pass
